### PR TITLE
[feat] chunkserver/datastore: add endianess in chunkfile/snapshotfile metapage

### PR DESCRIPTION
**Pull request by mOUNT41N**
### What problem does this PR solve?

Issue Number: #2260 

Problem Summary: 

ChunkFileMetaPage and SnapshotMetaPage does not take endianness into consideration

What's Changed:

Take endianness into consideration
Add verison 

Side effects(Breaking backward compatibility? Performance regression?):
The metapage's attributes will convert to network endianess after encode. 
Change metapage the version if the metapage is new or have been decoded.

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by mOUNT41N** [comment]
cicheck

**Comment by mOUNT41N** [comment]
cicheck

**Comment by zhanghuidinah** [comment]
cicheck

**Comment by mOUNT41N** [comment]
cicheck

File: src/chunkserver/datastore/chunkserver_chunkfile.cpp [hunk 1]
```diff
@@ -74,56 +75,130 @@ ChunkFileMetaPage& ChunkFileMetaPage::operator =(
...
+    uint64_t netCorrectedSn = htonll(correctedSn);
+    memcpy(buf + len, &netCorrectedSn, sizeof(netCorrectedSn));
+    len += sizeof(netCorrectedSn);
+
+    // long unsigned int loc_size 4/8 bytes need convert to network endianness
     size_t loc_size = location.size();
-    memcpy(buf + len, &loc_size, sizeof(loc_size));
-    len += sizeof(loc_size);
+    size_t net_loc_size;
+    if (sizeof(size_t) == 4) {
```
Comment (xu-chaojie) [inline src/chunkserver/datastore/chunkserver_chunkfile.cpp:110]: 不用考虑支持多平台，目前只考虑64位就可以了，支持多平台也不是这么支持的，到使用的时候去判断多少位是不合适的，判断不过来的

File: src/chunkserver/datastore/chunkserver_chunkfile.cpp [hunk 2]
```diff
@@ -74,56 +75,130 @@ ChunkFileMetaPage& ChunkFileMetaPage::operator =(
     return *this;
 }
 
-void ChunkFileMetaPage::encode(char* buf) {
+uint64_t ChunkFileMetaPage::htonll(uint64_t val) {
```
Comment (xu-chaojie) [inline src/chunkserver/datastore/chunkserver_chunkfile.cpp:78]: use the functions defined in the endian.h header file, or completely implement the endian conversion functions by youself.

**Comment by mOUNT41N** [comment]
cicheck

**Comment by mOUNT41N** [comment]
cicheck

**Comment by mOUNT41N** [comment]
cicheck

**Comment by mOUNT41N** [comment]
cicheck

**Comment by mOUNT41N** [comment]
cicheck

**Comment by mOUNT41N** [comment]
cicheck

**Comment by mOUNT41N** [comment]
cicheck

**Comment by mOUNT41N** [comment]
cicheck

**Comment by mOUNT41N** [comment]
recheck

**Comment by mOUNT41N** [comment]
> @mOUNT41N You see there is a segment fault in the unit test image http://59.111.91.248:8080/job/curve_untest_job/6553/console

ok, i'll find the reason.

**Comment by mOUNT41N** [comment]
cicheck

**Comment by mOUNT41N** [comment]
recheck

**Comment by mOUNT41N** [comment]
cicheck

**Comment by ilixiaocui** [comment]
@xu-chaojie Please take a look.

**Comment by ilixiaocui** [comment]
@wuhongsong Please deal with the conflict.

**Comment by ilixiaocui** [comment]
This PR has not been updated for a long time, may I ask where it is currently stuck?

 If the communication on the issue is not enough, you can join our developer group, please note the issue number when entering the group.
image


 If you have already joined, can you explain the WeChat ID?

**Comment by mOUNT41N** [comment]
> This PR has not been updated for a long time, may I ask where it is currently stuck?
> 
> If the communication on the issue is not enough, you can join our developer group, please note the issue number when entering the group. image
> 
> If you have already joined, can you explain the WeChat ID?

**Comment by mOUNT41N** [comment]
cicheck

**Comment by mOUNT41N** [comment]
When considering the crc int metapage, it shows the problem of verifing copyset status. But crc is not defined in ChunkFileMetaPage and SnapshotMetapage. And final I'm just consider the endianness of attributes that metapage defines. Though the checks have passed, it seems iops problems will happen somtime.

**Comment by ilixiaocui** [comment]
@xu-chaojie Please review this pr ASAP!!!

**Comment by Cyber-SiKu** [comment]
@xu-chaojie Please review this pr!

File: src/chunkserver/datastore/chunkserver_chunkfile.cpp [hunk 3]
```diff
@@ -74,74 +75,131 @@ ChunkFileMetaPage& ChunkFileMetaPage::operator =(
...
+        // bitmap char  (bits + 8 - 1) / 8 bytes
         size_t bitmapBytes = (bits + 8 - 1) >> 3;
         memcpy(buf + len, bitmap->GetBitmap(), bitmapBytes);
         len += bitmapBytes;
     }
+
+    // uint32_t crc 4 bytes
     uint32_t crc = ::curve::common::CRC32(buf, len);
-    memcpy(buf + len, &crc, sizeof(crc));
+    uint32_t beCrc = crc;
```
Comment (xu-chaojie) [inline src/chunkserver/datastore/chunkserver_chunkfile.cpp:122]: Why is crc not converted to big endian？

File: src/chunkserver/datastore/chunkserver_chunkfile.cpp [hunk 4]
```diff
@@ -74,74 +75,131 @@ ChunkFileMetaPage& ChunkFileMetaPage::operator =(
...
+    memcpy(buf + len, &beCrc, sizeof(beCrc));
+
+    LOG(INFO) << "calculate crc:" << crc;
 }
 
-CSErrorCode ChunkFileMetaPage::decode(const char* buf) {
+CSErrorCode ChunkFileMetaPage::decode(const char *buf) {
```
Comment (xu-chaojie) [inline src/chunkserver/datastore/chunkserver_chunkfile.cpp:128]: If you decode a FORMAT_VERSION_V2 version of buf, will it still work correctly?

File: src/chunkserver/datastore/chunkserver_snapshot.cpp [hunk 5]
```diff
@@ -20,102 +20,142 @@
...
+    // unsigned long bitmapBytes 4 bytes
+    // bitmap char  (bits + 8 - 1) / 8 bytes
+    size_t bitmapBytes = (bits + 8 - 1) >> 3;
     memcpy(buf + len, bitmap->GetBitmap(), bitmapBytes);
     len += bitmapBytes;
+
+    // uint32_t crc 4 bytes
     uint32_t crc = ::curve::common::CRC32(buf, len);
-    memcpy(buf + len, &crc, sizeof(crc));
+    uint32_t beCrc = crc;
```
Comment (xu-chaojie) [inline src/chunkserver/datastore/chunkserver_snapshot.cpp:61]: Why is crc not converted to big endian？

File: src/chunkserver/datastore/chunkserver_snapshot.cpp [hunk 6]
```diff
@@ -20,102 +20,142 @@
...
+    memcpy(buf + len, &beCrc, sizeof(beCrc));
+
+    LOG(INFO) << "calculate crc:" << crc;
 }
 
-CSErrorCode SnapshotMetaPage::decode(const char* buf) {
+CSErrorCode SnapshotMetaPage::decode(const char *buf) {
```
Comment (xu-chaojie) [inline src/chunkserver/datastore/chunkserver_snapshot.cpp:67]: same as ChunkFileMetaPage::decode

File: src/chunkserver/datastore/define.h [hunk 7]
```diff
@@ -34,10 +34,14 @@ namespace chunkserver {
 
 using curve::common::Bitmap;
 
+// In zeroed chunk file, taking endianness into consideration, the version is 8,
+// Taking endianness into consideration, the version is 4,
 // In zeroed chunk file, the version is 2,
 // otherwise, the version is 1
 const uint8_t FORMAT_VERSION = 1;
 const uint8_t FORMAT_VERSION_V2 = 2;
+const uint8_t FORMAT_VERSION_V3 = 4;
```
Comment (xu-chaojie) [inline src/chunkserver/datastore/define.h:43]: why not equal to 3 or 4 ？

File: src/chunkserver/datastore/define.h [hunk 8]
```diff
@@ -34,10 +34,14 @@ namespace chunkserver {
...
+const uint8_t FORMAT_VERSION_V4 = 8;
```
Comment (xu-chaojie) [inline src/chunkserver/datastore/define.h:44]: why not  is 3  and 4 ?

File: src/chunkserver/datastore/chunkserver_snapshot.cpp [hunk 9]
```diff
@@ -143,23 +183,21 @@ CSErrorCode CSSnapshot::Open(bool createFile) {
...
         int ret = chunkFilePool_->GetFile(snapshotPath, buf.get());
         if (ret != 0) {
             LOG(ERROR) << "Error occured when create snapshot."
-                   << " filepath = " << snapshotPath;
+                       << " filepath = " << snapshotPath;
             return CSErrorCode::InternalError;
         }
     }
-    int rc = lfs_->Open(snapshotPath, O_RDWR|O_NOATIME|O_DSYNC);
+    int rc = lfs_->Open(snapshotPath, O_RDWR | O_NOATIME | O_DSYNC);
```
Comment (xu-chaojie) [inline src/chunkserver/datastore/chunkserver_snapshot.cpp:197]: Do not introduce code style changes unrelated to the issue

**Comment by xu-chaojie** [comment]
rebase all commit to one commit，and fix conflict

File: src/chunkserver/datastore/chunkserver_chunkfile.cpp (hunk 3 above)
Comment (mOUNT41N) [reply src/chunkserver/datastore/chunkserver_chunkfile.cpp:122]: Beacue the crc is not defined in ChunkFileMetaPage and SnapshotMetapage.

File: src/chunkserver/datastore/chunkserver_snapshot.cpp (hunk 9 above)
Comment (mOUNT41N) [reply src/chunkserver/datastore/chunkserver_snapshot.cpp:197]: ok, i will fix it

File: src/chunkserver/datastore/define.h (hunk 8 above)
Comment (mOUNT41N) [reply src/chunkserver/datastore/define.h:44]: It's all right for 3 and 4, maybe I just set it as 2^(n-1) at that time, should i fix it?

**Comment by SeanHai** [comment]
@mOUNT41N  Are there any difficulties with this PR?

**Comment by Cyber-SiKu** [comment]
@mOUNT41N Are there any difficulties with this PR?

**Comment by mOUNT41N** [comment]
I'm sorry it wasn't finished in time , I'll fix the conflicts ASAP.

**Comment by Cyber-SiKu** [comment]
@xu-chaojie PTAL

**Comment by xu-chaojie** [comment]
Still not implemented correctly now @Cyber-SiKu

**Comment by mOUNT41N** [comment]
cicheck

**Comment by Cyber-SiKu** [comment]
pls fix conflicts.

**Comment by wuhongsong** [comment]
pr stale,you can reopen it later if you want continue.

**Pull request closed without merging**